### DISCUSSION TITLE
[4.6] OCI Getting location pop up

### DISCUSSION THREAD

**Discussion by ddevadat** [Sunbird-Ed/Community#69] [Installation] [answered]
We have deployed sunbird 4.6 on OCI.

Everytime we login we get a pop for location as given in the attachment.
Is there a way to disable this ?
<img width="1436" alt="Screen Shot 2022-08-09 at 10 23 07 am" src="https://user-images.githubusercontent.com/27432927/183536644-fe113a2b-42a3-40ee-8aed-6b7821c88e79.png">

**Comment by sharathkashyap** [comment]
@ddevadat This is primarily due  to device register api not up and running. Kindly see to it this service returns 200 status code. If the service is working fine. It might require some investigation to be done. Kindly come back on this and let us know.
cc: @luckynvdu

**Comment by ddevadat** [comment]
hi sharath,

how to check device register api status..is it a k8s deployment?

**Reply by sharathkashyap** [reply]
@ddevadat as part of initial check you can try to see the api response returning in browser alone. If it turns to be response status other than 200. We need to check if the API Manager has onboarded the API. Will get you some help in this regard.

cc: @keshavprasadms @luckynvdu

**Reply by sharathkashyap** [reply]
@ddevadat any updates on the thread.

**Comment by ddevadat** [comment]
This issue is still there ...i tried the echo api and its returning the value


`
curl --location --request GET 'https://XXXX.bird.co.in/api/echo/Howdy' \
--header 'Content-type: application/json' \
--header 'Authorization: Bearer eyXXXXXXyJpc3MiOiJiOGliRkRRM3p6T0hVUnJURkk0ZW9ia2FDV2wxaEt1UCJ9.ffffJuk' \
--data-raw '{

    "request": 

{

    "channel": 

        {
            "name": "Test Channel",
            "code": "testchannelid0001"
        }
    }

}'

`

**Comment by ddevadat** [comment]
for the player pod config map..there is a variable like

sunbird_device_register_api: https://api.XXXXX.ocisunbird.co.in/v3/device/register/

is this the device api url you are referring to ..

i tried changing it to
sunbird_device_register_api: https://kong:8000/v3/device/register/

But it made no difference

**Comment by ddevadat** [comment]
any update on this?

**Comment by ddevadat** [comment]
api calls for device registration is completing sucessfully , but after that on trying to retrieve its not returning anything

```
curl --location --request GET 'https://XXXXXX.XX.XXXX.co.in/api/v3/device/profile/ea1e3a365af53a8172fdecae0e5b447c' \
--header 'Content-type: application/json' \
--header 'Authorization: Bearer XXXX.eyJpc3MiOiJiOGliRkRRM3p6T0hVUnJURkk0ZW9ia2FDV2wxaEt1UCJ9.z2e09u4ArX7_U4Fq2ONtzRRi6JslCM6sXKZAORp8Juk' \
--data-raw ''

```


```
{
    "id": "analytics.device-profile",
    "ver": "1.0",
    "ts": "2022-09-21T04:17:11.276+00:00",
    "params": {
        "resmsgid": "7eabb8f0-46bd-42e1-9ecf-fac783f23e9e",
        "status": "successful",
        "client_key": null
    },
    "responseCode": "OK",
    "result": {
        "ipLocation": {
            "state": "",
            "district": ""
        }
    }
}

```


its always returns empty values for state and districts

**Reply by beepdot** [reply]
Hello @ddevadat Let us try to debug the issue. Let's try to see if we are able to see the public ip first in the logs. 
1. When you invoke the device profile API in curl / postman, please tail the `nginx-public-ingress` pod logs and post the output here. It would look something like this
```
106.51.75.240 - - [21/Sep/2022:04:45:22 +0000] "GET /device/profile/f1bf225bd30df9cbda9695421fc194a2 HTTP/1.1" 200 910 252 0.027 0.024 . "https://dev.sunbirded.org/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36" "a68c4556-0ce4-5361-e982-1527cbf55aaa" "f1bf225bd30df9cbda9695421fc194a2" "-" "dev.sunbird.portal" "-" "a1ce5a6d-a0cb-51ce-1235-687945041832" 
```
2. Please invoke the API again and this this time tail the logs of `apimanager` pods and post the output here
```
106.51.75.240 - - [21/Sep/2022:04:43:12 +0000] "GET /v3/device/profile/f1bf225bd30df9cbda9695421fc194a2 HTTP/1.1" 200 731 284 0.020 0.008 . "-" "-" "-" "-" "-" "-" "-" "-"
```
Based on the output you provide me, I will be able to ask you additional information.

**Comment by ddevadat** [comment]
from nginx-public-ingress

```
10.0.1.115 - - [21/Sep/2022:04:57:10 +0000] "GET /api/v3/device/profile/ea1e3a365af53a8172fdecae0e5b447c HTTP/1.1" 200 459 254 0.018 0.018 . "-" "PostmanRuntime/7.29.2" "b2c477d97a089f29eeb4cf7af4cedd62" "-" "-" "-" "-" "-"

```

from apimanager pod

```
10.0.1.115 - - [21/Sep/2022:04:59:31 +0000] "GET /v3/device/profile/ea1e3a365af53a8172fdecae0e5b447c HTTP/1.1" 200 548 254 0.016 0.006 . "-" "PostmanRuntime/7.29.2" "b592a1b2370b3b8992bfafc1ac2b0ddd" "-" "-" "-" "-" "-"
```

**Reply by beepdot** [reply]
Is this your public ip - `10.0.1.115`? Please check your public ip and let me know - https://whatismyipaddress.com/

**Comment by ddevadat** [comment]
its not my public ip...i tried invoking from the jenkins server using curl and i get the same ip

```
10.0.1.115 - - [21/Sep/2022:05:09:54 +0000] "GET /api/v3/device/profile/ea1e3a365af53a8172fdecae0e5b447c HTTP/1.1" 200 355 254 0.098 0.098 . "-" "curl/7.58.0" "1def5bb2e9a287486966fe6c6f2d0f67" "-" "-" "-" "-" "-"

```

**Reply by beepdot** [reply]
Can you please post me the output of the below command? Please add your respective kubernetes namespace to below commad.
```
kubectl get svc nginx-public-ingress -ojsonpath='{.spec.externalTrafficPolicy}'
```

**Comment by ddevadat** [comment]
```
kubectl get svc nginx-public-ingress -ojsonpath='{.spec.externalTrafficPolicy}'
Local

```

**Reply by beepdot** [reply]
Ok, So thats one part of the issue. Lets check if we can retrieve the location from the analytics pods. Login to the analytics pod and run the below command. Please share the output. Please replace the device id and IP address as per your data.
```
curl -H 'X-Real-IP: 106.51.75.240' localhost:9000/v1/device/profile/f1bf225bd30df9cbda9695421fc194a2
```
The output should look like this.
```
{"id":"analytics.device-profile","ver":"1.0","ts":"2022-09-21T05:47:49.281+00:00","params":{"resmsgid":"d02191b7-e54e-4784-8e4d-560f65b20344","status":"successful","client_key":null},"responseCode":"OK","result":{"ipLocation":{"state":"Karnataka","district":"BENGALURU URBAN SOUTH"}}}
```

**Comment by ddevadat** [comment]
Executed from analytics pod

```
/home/analytics $ curl -H 'X-Real-IP: 10.0.1.115' localhost:9000/v1/device/profile/ea1e3a365af53a8172fdecae0e5b447c
{"id":"analytics.device-profile","ver":"1.0","ts":"2022-09-21T06:01:21.009+00:00","params":{"resmsgid":"20cca50e-78da-4ce6-9ead-c47750cfec19","status":"successful","client_key":null},"responseCode":"OK","result":{"ipLocation":{"state":"","district":""}}}

```

**Reply by beepdot** [reply]
Please use your real public IP address instead of `10.0.1.115`

**Comment by ddevadat** [comment]
i used the public ip of jenkins server

```
/home/analytics $ curl -H 'X-Real-IP: 140.238.242.190' localhost:9000/v1/device/profile/ea1e3a365af53a8172fdecae0e5b447c
{"id":"analytics.device-profile","ver":"1.0","ts":"2022-09-21T06:17:04.250+00:00","params":{"resmsgid":"493db681-1b14-4aa0-aa39-a4ac53a988c9","status":"successful","client_key":null},"responseCode":"OK","result":{"ipLocation":{"state":"Maharashtra","district":"Mumbai"}}}/home/analytics $

```

**Reply by beepdot** [reply]
I am not aware of OCI cloud. You may want to read the docs here - https://docs.oracle.com/en-us/iaas/Content/ContEng/Tasks/contengcreatingloadbalancer.htm#contengcreatingnetworkloadbalancer_topic_Preserving_client_IP

What we require is the original IP address coming in nginx logs. The services and configuration from Sunbird end look fine. Please check how to ensure the client ip is sent to nginx pods from the load balancers. Please do post back if you are having issues and also in case you find a solution.

Once nginx gets the original client's IP address, it will be forwarded to API manager and then to Analytics pods. The analytics pods are able to resolve a location based on the IP from the output you shared above.
`{"ipLocation":{"state":"Maharashtra","district":"Mumbai"}}`

**Reply by beepdot** [reply]
You can also check if you need to modify any headers if required here - https://github.com/project-sunbird/sunbird-devops/blob/master/kubernetes/helm_charts/core/nginx-public-ingress/values.j2#L856-L869
This is the header we are interested in
`proxy_set_header X-Real-IP {{ nginx_client_public_ip_header | d('$remote_addr') }};`

**Comment by ddevadat** [comment]
do we need to set this header X-Real-IP for all location or only 
location /

Also where is the value of nginx_client_public_ip_header defined

**Reply by beepdot** [reply]
For the device register API, only in `location /` its sufficient. But all location blocks have this. If you are interested to capture the client's ip address for various URLs, then you should modify across locations. So its your choice if you want to make changes in all locations or just at `/`.
You can ignore `nginx_client_public_ip_header` and just use `$remote_addr`.  This would be the same even in your configmap named `proxy-default` where it would be using `$remote_addr`.

**Comment by ddevadat** [comment]
i didnt see any specific flag for preserving client ip in public load balancer for oci.
just curious is there any flag we set if this was deployed in azure

**Reply by beepdot** [reply]
Nope. In azure cloud we dont need to set anything at LB level.  For Kubernetes, we just need to set to `Local` which you already have done.
https://github.com/orgs/Sunbird-Ed/discussions/69#discussioncomment-3696557

**Reply by gandham-santhosh** [reply]
@ddevadat please check and mark this thread

**Comment by ddevadat** [comment, marked as answer]
this is not longer needed
